### PR TITLE
[inductor] fix cpp legalize bf16 reduction

### PR DESCRIPTION
**Pull request by Valentine233**
When legalizing bf16 for reduction, operators with result dtype of torch.int64, like argmax, would encounter an assertion error now. The PR fixes for the case of int64, enabling several bf16 models (hf_Reformer, doctr_reco_predictor) to run successfully.

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_inductor/codegen/cpp.py [hunk 1]
```diff
@@ -2037,16 +2037,27 @@ def add_to_dtype(sub_graph: torch.fx.Graph):
                         # Since we always convert the load/store value to float if the tensor is bfloat16.
                         # Therefore, the reduction should never work with bfloat16 value. Hence, we update
                         # the bfloat16 reduction by updating the dtype and src_dtype to float.
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp.py:2039]: Can you update the comments too?

**Review by jgong5 — APPROVED** [review]

**Review by EikanWang — APPROVED** [review]

File: torch/_inductor/codegen/cpp.py (hunk 1 above)
Comment (Valentine233) [reply torch/_inductor/codegen/cpp.py]: Updated.

**Review by EikanWang — APPROVED** [review]

File: torch/_inductor/codegen/cpp.py [hunk 2]
```diff
@@ -2036,17 +2036,30 @@ def add_to_dtype(sub_graph: torch.fx.Graph):
...
-                        )
+                        # the bfloat16 reduction by
+                        #    1) updating the src_dtype to float if dtype is int64
+                        # or 2) updating the dtype and src_dtype to float if else.
+                        assert dtype in [torch.float, torch.bfloat16, torch.int64]
+                        if dtype == torch.int64:
+                            _node.args = (
+                                ops,
+                                name,
+                                torch.int64,
```
Comment (desertfire) [inline torch/_inductor/codegen/cpp.py:2047]: We could fold the if-condition to here if this is the only difference between the two branches.

**Review by desertfire — APPROVED** [review]

File: torch/_inductor/codegen/cpp.py (hunk 2 above)
Comment (Valentine233) [reply torch/_inductor/codegen/cpp.py]: Thanks. It seems much cleaner.

**Pull request closed without merging**